adam/learner/propose_but_verify.py

```diff
@@ -0,0 +1,372 @@
+    _observation_num = attrib(init=False, default=0)
+    _rng: Random = attrib(validator=instance_of(Random))
+    _words_to_hypotheses: Dict[str, PerceptionGraphPattern] = attrib(
+        init=False, default=Factory(dict)
+    )
+    _log_word_hypotheses_to: Optional[Path] = attrib(
+        validator=optional(instance_of(Path)), default=None
+    )
+    _graph_logger: Optional[GraphLogger] = attrib(
+        validator=optional(instance_of(GraphLogger)), default=None
+    )
+    debug_counter = 0
+    _debug_callback: Optional[DebugCallableType] = attrib(default=None)
+    _graph_match_confirmation_threshold: float = attrib(default=0.9, kw_only=True)
+    _word_to_logger: Dict[str, GraphLogger] = attrib(init=False, default=Factory(dict))
+    _rendered_word_hypothesis_pair_ids: Set[str] = attrib(
+        init=False, default=Factory(set)
+    )
+
+    @staticmethod
+    def from_parameters(
+        params: Parameters, *, graph_logger: Optional[GraphLogger] = None
+    ) -> "ProposeButVerifyLanguageLearner":  # type: ignore
+        log_word_hypotheses_dir = params.optional_creatable_directory(
+            "log_word_hypotheses_dir"
+        )
+        if log_word_hypotheses_dir:
+            logging.info("Hypotheses will be logged to %s", log_word_hypotheses_dir)
+
+        rng = Random()
+        rng.seed(params.optional_integer("random_seed", default=0))
+
+        return ProposeButVerifyLanguageLearner(
+            graph_match_confirmation_threshold=params.floating_point(
+                "graph_match_confirmation_threshold"
+            ),
+            graph_logger=graph_logger,
+            log_word_hypotheses_to=log_word_hypotheses_dir,
+            rng=rng,
+            ontology=GAILA_PHASE_1_ONTOLOGY,
+        )
+
+    # observe method
+    def observe(
+        self, learning_example: LearningExample[PerceptionT, LinguisticDescription]
+    ) -> None:
+
+        logging.info("Observation %s", self._observation_num)
+        self._observation_num += 1
+
+        perception = learning_example.perception
+        if len(perception.frames) != 1:
+            raise RuntimeError("Pursuit learner can only handle single frames for now")
+        if isinstance(perception.frames[0], DevelopmentalPrimitivePerceptionFrame):
+            original_perception_graph = PerceptionGraph.from_frame(
+                perception.frames[0]
+            ).copy_as_digraph()
+        else:
+            raise RuntimeError("Cannot process perception type.")
+        # Remove learner from the perception
+        observed_perception_graph = graph_without_learner(original_perception_graph)
+        observed_linguistic_description = (
+            learning_example.linguistic_description.as_token_sequence()
+        )
+
+        self.learn_propose_but_verify(
+            observed_perception_graph, observed_linguistic_description
+        )
+
+    # learn method
+    def learn_propose_but_verify(
```

Comment:
Can you add tests that checks if this is working as intended? Something similar to PursuitLearner tests with single-object situations would be enough

Reply:
Tests have been added to `learner_tests`. It still uses the pursuit curriculum since it works as intended.